File: Game Client/game/Menu.py
from .config import *
from .Player import *
from .Wall import *
from .Ball import *
from .StartScreen import *
from .WaitScreen import *
from .Pause import *
from .AI import *
from .Custom import *
from .Button import *
from .TournamentMenu import *

class Menu:
	def __init__(self):
		self.title_font = pg.font.Font(font, int(winHeight / 3))
		self.text_font = pg.font.Font(font, int(winHeight * 0.05))
		self.button_size = [winWidth * 0.1, winHeight * 0.1]
		self.buttons = [Button("SOLO", (winWidth / 3) - (self.button_size[0] / 2), (winHeight / 2) - self.button_size[1], self.button_size[0], self.button_size[1], winHeight * 0.085),
					Button("LOCAL", (winWidth / 3 * 2) - (self.button_size[0] / 2), (winHeight / 2) - self.button_size[1], self.button_size[0], self.button_size[1], winHeight * 0.085),
					Button("ONLINE", (winWidth / 3) - (self.button_size[0] / 2), (winHeight / 3 * 2), self.button_size[0], self.button_size[1], winHeight * 0.085),
					Button("CUSTOM", (winWidth / 3 * 2) - (self.button_size[0] / 2), (winHeight / 3 * 2), self.button_size[0], self.button_size[1], winHeight * 0.085),
					Button("JOIN", 25, winHeight - self.button_size[1] - 25, self.button_size[0], self.button_size[1], winHeight * 0.085),
					Button("", self.button_size[0] + 35, winHeight - self.button_size[1] - 25, self.button_size[0], self.button_size[1], winHeight * 0.085),
	  				Button("TOURNAMENT", winWidth * 0.815, winHeight * 0.875, self.button_size[0] * 1.75, self.button_size[1], winHeight * 0.085)]
		self.err = False
  
	def draw(self, win):
		title = self.title_font.render("PONG", True, (255, 255, 255))
		win.blit(title, ((winWidth / 2) - (title.get_size()[0] / 2), -50))
		if self.err:
			text = self.text_font.render(self.err, True, (255, 102, 102))
			win.blit(text, (self.buttons[5].x + self.button_size[0] + 15, self.buttons[5].y + (self.button_size[1] * 0.2)))
  
		for button in self.buttons:
			button.draw(win)
   
   
	async def click(self, core, mousePos):
		for button in self.buttons:
			if button.hitbox.collidepoint(mousePos):
				await self.setValues(button.name, core)
	
 
	async def setValues(self, name, core):
		core.customs = []
		core.square = False
		core.obstacle = False
		if name == 'JOIN':
			if self.buttons[5].name.__len__() == 0:
				self.err = "Room id is empty"
				return
			await core.GameHub.send(json.dumps({'type' : 'join', 'id' : self.buttons[5].name}))
			response : dict = json.loads(await core.GameHub.recv())
			if response['success'] == 'false':
				if "error" in response.keys():
					self.err = response['error']
				else:
					self.err = "Room " + self.buttons[5].name + " doesn't exist"
			else:
				core.GamePort = response['port']
				room_id = self.buttons[5].name
				core.id = response['pos']
				core.mode = "ONLINE"
				core.online = True
				if response['mode'] == 'tournament':
					core.tournament = Tournament(response['custom_mods'], response['max'], response['ai'], response['score'], True, False)
					core.players = []
					for i in range(response['players'].__len__()):
						core.players.append(Player(i + 1, response['players'][i], 2, 'BORDERLESS' in response['custom_mods'], False))
					await core.tournament.initPlayers(core.players)
					core.tournament.id = room_id
					core.tournament_id = core.id
					core.state = 'tournament'
				else:
					core.state = "waiting"
					wait_nb = response['max']
					core.square = "1V1V1V1" in response['custom_mods']
					core.start_screen = StartScreen(response['mode'], core.online, core.square, wait_nb)
				core.customs = response['custom_mods'] if response['custom_mods'].__len__() > 0 else False

		if name == self.buttons[5].name:
			self.buttons[5].highlight = not self.buttons[5].highlight

		if name == "LOCAL":
			msg = {"type" : "quickGame", "cmd" : "join", "online" : "false"}
			await core.GameHub.send(json.dumps(msg))
	
			core.players = [Player(1, core.alias + "1", 2, False, False), Player(2, core.alias + "2", 2, False, False)]
			core.walls = [Wall("up", False), Wall("down", False)]
			core.ball = Ball(False)
			core.state = "start"
			core.mode = "LOCAL"
	
		if name == "SOLO":
			msg = {"type" : "quickGame", "cmd" : "join", "online" : "false"}
			await core.GameHub.send(json.dumps(msg))
	
			core.players = [Player(1, core.alias, 2, False, False), Player(2, "AI", 2, False, False)]
			core.ai.append(AI(core.players[1]))
			core.walls = [Wall("up", False), Wall("down", False)]
			core.ball = Ball(False)
			core.state = "start"
			core.mode = "solo"
	
		if name == "CUSTOM":
			core.state = "custom"
			core.custom_menu = CustomMenu()
			self.buttons[5].name = ''
			self.buttons[5].highlight = False
			self.err = False
		
		if name == "ONLINE":
			msg = {"type" : "quickGame", "cmd" : "join", "online" : "true"}
			await core.GameHub.send(json.dumps(msg))
			response : dict = json.loads(await core.GameHub.recv())
			if 'port' in response.keys():
				core.GamePort = response['port']
				room_id = response['ID']
				core.state = "waiting"
				core.mode = "ONLINE"
				core.id = response['pos']	
			core.online = True
			wait_nb = 2

		if name == "TOURNAMENT":
			core.tournament_menu = TournamentMenu()
			core.state = 'tournament menu'
			self.buttons[5].name = ''
			self.buttons[5].highlight = False
			self.err = False

		if core.mode != "none":
			self.buttons[5].name = ""
			self.buttons[5].highlight = False
			self.err = False
			if not core.start_screen:
				core.start_screen = StartScreen(core.mode, core.online)
			if core.online and not core.wait_screen and core.state != 'tournament':
				core.wait_screen = WaitScreen(room_id, core.id, wait_nb, "QuickGame Online")

File: srcs/django_code/transcend/myapp/static/js/game/Menu.js
import { canvas, ctx } from "./canvas.js";
import { Player } from "./Player.js";
import { Wall } from "./Wall.js";
import { Ball } from "./Ball.js";
import { StartScreen } from "./StartScreen.js";
import { WaitScreen } from "./WaitScreen.js";
import { Pause } from "./Pause.js";
import { AI } from "./AI.js";
import { CustomMenu } from "./Custom.js";
import { Button } from "./Button.js";
import { Vec2 } from "./Vec2.js";
import { is_colliding } from "./Hitbox.js";

export class Menu {
	constructor() {
		this.button_size = [canvas.width * 0.1, canvas.height * 0.1];
		this.buttons = [new Button("SOLO", (canvas.width / 3) - (this.button_size[0] / 2), (canvas.height / 2) - this.button_size[1], this.button_size[0], this.button_size[1]),
				  new Button("LOCAL", (canvas.width / 3 * 2) - (this.button_size[0] / 2), (canvas.height / 2) - this.button_size[1], this.button_size[0], this.button_size[1]),
				  new Button("ONLINE", (canvas.width / 3) - (this.button_size[0] / 2), (canvas.height / 3 * 2), this.button_size[0], this.button_size[1]),
				  new Button("CUSTOM", (canvas.width / 3 * 2) - (this.button_size[0] / 2), (canvas.height / 3 * 2), this.button_size[0], this.button_size[1]),
				  new Button("JOIN", (canvas.width * 0.01), (canvas.height * 0.875), this.button_size[0], this.button_size[1]),
				  new Button("", (canvas.width * 0.12), (canvas.height * 0.875), this.button_size[0], this.button_size[1])];
		this.err = false;
	}

	draw() {
		ctx.font = Math.floor(canvas.height / 3) + "px pong-teko";
		ctx.fillText("PONG", canvas.width / 2, canvas.height / 4);
		ctx.font = Math.floor(canvas.height * 0.05) + "px pong-teko";
		if (this.err) {
			ctx.fillStyle = "rgb(255, 102, 102)";
			ctx.fillText(this.err, this.buttons[5].x + this.button_size[0] * 1.7, this.buttons[5].y + this.button_size[1] * 0.55);
			ctx.fillStyle = "white";
		}
		ctx.font = Math.floor(canvas.height * 0.085) + "px pong-teko";
		for (let b of this.buttons)
			b.draw();
	}

	click(core, mousePos) {
		const pos = new Vec2(mousePos[0], mousePos[1]);
		for (let b of this.buttons)
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.button_size))
				this.setValues(b.name, core);
	}

	setValues(name, core) {
		core.custom_mod = false;
		core.obstacle = false;
		if (name === "JOIN") {
			if (this.buttons[5].name.length === 0) {
				this.err = "Room id is empty";
				return;
			}
			// await core.GameHub.send(json.dumps({'type' : 'join', 'id' : self.buttons[5].name}))
			// response : dict = json.loads(await core.GameHub.recv())
			if (response['success'] == 'false')
				this.err = "Room " + this.buttons[5].name + " doesn't exist";
			else {
				// core.GameSocket = response['socket']
				room_id = this.buttons[5].name;
				core.id = response['pos'];
				core.state = "waiting";
				core.mode = "ONLINE";
				core.online = true;
				wait_nb = response['max'];
				core.custom_mod = (response['custom_mods'].includes("1V1V1V1")) ? "1V1V1V1" : false;
				core.start_screen = new StartScreen(response['mode'], core.online, response['custom_mods'].includes("1V1V1V1"), wait_nb);
			}
		}
		if (name === this.buttons[5].name)
			this.buttons[5].highlight = !this.buttons[5].highlight;
		if (name === "LOCAL") {
			// msg = {"type" : "quickGame", "cmd" : "join", "online" : "false"}
			// await core.GameHub.send(json.dumps(msg))
			core.players = [new Player(1, core.alias + "1", 2, false, false), new Player(2, core.alias + "2", 2, false, false)];
			core.walls = [new Wall("up", false), new Wall("down", false)];
			core.ball = new Ball(false);
			core.state = "start";
			core.mode = "LOCAL";
		}
		if (name === "SOLO") {
			// msg = {"type" : "quickGame", "cmd" : "join", "online" : "false"}
			// await core.GameHub.send(json.dumps(msg))
			core.players = [new Player(1, core.alias, 2, false, false), new Player(2, "AI", 2, false, false)];
			core.ai.push(new AI(core.players[1]));
			core.walls = [new Wall("up", false), new Wall("down", false)];
			core.ball = new Ball(false);
			core.state = "start";
			core.mode = "solo";
		}
		if (name === "CUSTOM") {
			core.custom_menu = new CustomMenu();
			core.state = "custom";
		}
		if (name === "ONLINE") {
			// msg = {"type" : "quickGame", "cmd" : "join", "online" : "true"}
			// await core.GameHub.send(json.dumps(msg))
			// response : dict = json.loads(await core.GameHub.recv())
			if ("socket" in response) {
				core.GameSocket = response['socket'];
				room_id = response['ID'];
				core.mode = "ONLINE";
				core.state = "waiting";
				core.id = response['pos'];
			}
			core.online = true;
			wait_nb = 2;
		}
		if (core.mode != "none") {
			this.buttons[5].name = "";
			this.err = false;
			if (!core.start_screen)
				core.start_screen = new StartScreen(core.mode, core.online);
			if (core.online && !core.wait_screen)
				core.wait_screen = new WaitScreen(room_id, core.id, wait_nb, "QuickGame Online");
		}
	}
}